graphlearn_torch/python/distributed/dist_feature.py

```diff
@@ -203,6 +236,142 @@ def _local_selecting_get(
+        partition_ids = pb[ids.to(pb.device)].to(self.device)
+        send_remote_count = []
+        for pidx in range(0, self.num_partitions):
+          if pidx == self.partition_idx:
+            send_remote_count.append(0)
+          else:
+            remote_mask = (partition_ids == pidx)
+            remote_ids = torch.masked_select(ids, remote_mask)
+            ssize = remote_ids.numel()
+            send_remote_count.append(ssize)
+      send_sr, recv_sr = communicate_node_num(send_remote_count)
+      rfeat_recv_dict[ntype] = sum(recv_sr)
+      rfeat_send_dict[ntype] = sum(send_sr)
+    return rfeat_send_dict, rfeat_recv_dict
+
+  def communicate_node_id (
+    self,
+    sampler_result: Union[SamplerOutput, HeteroSamplerOutput],
+    ntype_list: List[NodeType]
+  ):
+    offset = 0
+    indexes = {}
+    send_ids = []
+    remote_cnt_list = torch.zeros(self.num_partitions, dtype=torch.long)
+    for ntype in ntype_list:
+      indexes[ntype] = [None] * self.num_partitions
+    for pidx in range(0, self.num_partitions):
+      remote_cnt_sum = 0
+      for ntype in ntype_list:
+        nodes = sampler_result.node.get(ntype, None)
+        if nodes is None:
+          continue
+        nodes = nodes.to(torch.long)
+        _, pb = self._get_local_store(ntype)
+        input_order= torch.arange(nodes.size(0),
+                                  dtype=torch.long,
+                                  device=self.device)
+        partition_ids = pb[nodes.to(pb.device)].to(self.device)
+        nodes = nodes.to(self.device)
+        if pidx == self.partition_idx:
+          continue
+        else:
+          remote_mask = (partition_ids == pidx)
+          remote_ids = torch.masked_select(nodes, remote_mask)
+          indexes[ntype][pidx] = torch.masked_select(input_order, remote_mask)
+          ssize = remote_ids.numel()
+          send_ids[offset: offset + ssize] = remote_ids.tolist()
+          remote_cnt_sum = remote_cnt_sum + remote_ids.numel()
+          offset = offset + ssize
+      remote_cnt_list[pidx] = remote_cnt_sum
+
+    assert len(send_ids) == sum(remote_cnt_list)
+    send_sr, recv_sr = communicate_node_num(remote_cnt_list)
+    _, trecv = sum(send_sr), sum(recv_sr)
+    self.recv_rn_count = []
+    for pidx in range(self.num_partitions):
+      self.recv_rn_count.append(int(recv_sr[pidx]))
+    self.recv_rn_gnid = torch.zeros(trecv, dtype=torch.long)
+    dist.all_to_all_single(self.recv_rn_gnid, torch.tensor(send_ids),
+                           self.recv_rn_count, remote_cnt_list.tolist(),
+                           async_op=False)
+    return remote_cnt_list, indexes
+
+  def communicate_node_feats(
+    self,
+    ntype_list: List[NodeType],
+    remote_cnt: torch.Tensor,
+    send_num_dict: Dict[NodeType, int],
+    recv_num_dict: Dict[NodeType, int],
+    indexes: Dict[NodeType, List]
+  ):
+    rfeats_list = []
+    offset = 0
+    for ntype in ntype_list:
+      feat_num = recv_num_dict.get(ntype)
```

Comment:
Why different partitions are using the same `recv_num_dict`? It seems `recv_num_dict` stores the global feature sum of a node type that a partition should receive from the others.

Reply:
Here each rank with separated partition data will maintain an independent `recv_num_dict` variable, which records how many neighbor nodes' features need to be received from other ranks in total for each node type in this rank. For each rank, this `recv_num_dict` is different.

Reply:
Sorry I pasted my question on the wrong line number. The question I intended to asked is here:
https://github.com/alibaba/graphlearn-for-pytorch/pull/134/files#r1579295074